File: src/main/kotlin/com/advent2022/puzzle8/Puzzle8.kt
package com.advent2022.puzzle8

import com.advent2021.base.Base
import java.lang.Integer.max

data class Cell(
    val num: Int,
    val r: Int,
    val c: Int,
    var visible: Boolean = false
)
class Data {
    var grid = ArrayList<List<Cell>>()

    fun addRow(row: String) {
        grid.add(row.toCharArray().mapIndexed { index, it -> Cell(it - '0', grid.size, index) })
    }

    val numRows: Int
        get() = grid.size
    val numCols: Int
        get() = if (grid.isEmpty()) 0 else grid.first().size
    fun cell(r: Int, c: Int) = grid[r][c]
    fun cellOrNull(r: Int, c: Int) = if (hasCell(r, c)) cell(r, c) else null
    fun hasCell(r: Int, c: Int) = r in 0 until numRows && c in 0 until numCols
    val allCells: List<Cell>
        get() = grid.flatten()

    fun scenicScore(cell: Cell): Int {
        return canSeeLeft(cell) * canSeeUp(cell) * canSeeRight(cell) * canSeeDown(cell)
    }

    private fun canSee(cell: Cell, fn: (cell: Cell) -> Cell?): Int {
        var nextCell: Cell? = fn(cell)
        var ret = 0
        while (nextCell != null) {
            ++ret
            if (nextCell.num < cell.num) {
                nextCell = fn(nextCell)
            } else {
                break
            }
        }
        return ret
    }

    private fun canSeeLeft(cell: Cell): Int = canSee(cell) { cellOrNull(it.r, it.c - 1) }
    private fun canSeeRight(cell: Cell): Int  = canSee(cell) { cellOrNull(it.r, it.c + 1) }
    private fun canSeeUp(cell: Cell): Int = canSee(cell) { cellOrNull(it.r - 1, it.c) }
    private fun canSeeDown(cell: Cell): Int  = canSee(cell) { cellOrNull(it.r + 1, it.c) }
}
typealias Solution = Int
typealias Solution2 = Solution

fun main() {
    try {
        val puz = Puzzle8()
        val solution1 = puz.solvePuzzle("inputsTest.txt", Data())
        println("Solution1: $solution1")

        val solution2 = puz.solvePuzzle2("inputs.txt", Data())
        println("Solution2: $solution2")
    } catch (t: Throwable) {
        t.printStackTrace()
    }
}
fun nullMax(a: Int?, b: Int?): Int? = if (a == null) b else if (b == null) a else max(a, b)
fun visible(num: Int, max: Int?) = num > (max ?: -1)

class Puzzle8 : Base<Data, Solution?, Solution2?>() {
    override fun parseLine(line: String, data: Data) {
        data.addRow(line)
    }

    override fun computeSolution(data: Data): Solution {
        computeVisible(data)
        return data.grid.sumOf { row -> row.count { it.visible } }
    }

    private fun computeVisible(data: Data) {
        for (r in 0 until data.numRows) {
            // forward
            var maxLeft: Int? = null
            for (c in 0 until data.numCols) {
                val cell = data.cell(r, c)
                cell.visible = cell.visible || visible(cell.num, maxLeft)
                maxLeft = nullMax(cell.num, maxLeft)
            }

            // backwards
            var maxRight: Int? = null
            for (c in data.numCols - 1 downTo 0) {
                val cell = data.cell(r, c)
                cell.visible = cell.visible || visible(cell.num, maxRight)
                maxRight = nullMax(cell.num, maxRight)
            }

        }
        for (c in 0 until data.numCols) {
            // down
            var maxUp: Int? = null
            for (r in 0 until data.numRows) {
                val cell = data.cell(r, c)
                cell.visible = cell.visible || visible(cell.num, maxUp)
                maxUp = nullMax(cell.num, maxUp)
            }

            // up
            var maxDown: Int? = null
            for (r in data.numRows - 1 downTo 0) {
                val cell = data.cell(r, c)
                cell.visible = cell.visible || visible(cell.num, maxDown)
                maxDown = nullMax(cell.num, maxDown)
            }
        }
    }

    override fun computeSolution2(data: Data): Solution2 {
        computeVisible(data)
        val solution = data.allCells.maxOf { data.scenicScore(it) }
//        data.allCells.forEach { println("$it ${data.scenicScore(it)}") }
//        val cell = data.allCells.find { it.scenicScore == solution }
//        println(cell)
        return solution
    }
}

File: src/main/python/advent2024/Puzzle8.py
from Base import Base
from typing import Tuple

Point = Tuple[int, int]

class Grid:
  def __init__(self):
    self.num_rows = 0
    self.num_cols = 0
    self.antennae = dict[chr, list(Point)]()

  def add_antenna(self, antenna: Point, ch: chr):
    if self.antennae.get(ch, None) is None:
      self.antennae[ch] = list()
    self.antennae[ch].append(antenna)

  def get_antenna(self, point: Point) -> chr:
    for antenna in self.antennae:
      if point in self.antennae[antenna]:
        return antenna
    return ''

  def print_graph(self, antinodes: set[Point]):
    for row in range(self.num_rows):
      for col in range(self.num_cols):
        antenna = self.get_antenna((row, col))
        if antenna != '':
          print(antenna, end='')
        elif (row, col) in antinodes:
          print('#', end='')
        else:
          print('.', end='')
      print()

MyDataType = Grid

class Puzzle(Base[MyDataType, int, int]):
    def main(self):
      # 467 too high
      solution1: int = self.solve_puzzle("puzzle8/inputs.txt", MyDataType())  # type: ignore
      print("Solution1: {}".format(solution1))

      solution2: int = self.solve_puzzle2("puzzle8/inputsTest.txt", MyDataType()) # type: ignore
      print("Solution2: {}".format(solution2))

    def read_one_line(self, line: str, data: MyDataType):
      data.num_cols = len(line)
      for col in range(data.num_cols):
        if line[col] != '.':
          pos = data.num_rows, col
          data.add_antenna(pos, line[col])
      data.num_rows += 1

    def compute_solution(self, data: MyDataType) -> int:
      antinodes = set[Point]()
      for ch in data.antennae:
        points = data.antennae[ch]
        for i in range(len(points) - 1):
          for j in range(i + 1, len(points)):
            antinodes |= self.calculate_antinodes(points[i], points[j], data)

      data.print_graph(antinodes)

      return len(antinodes)

    def compute_solution2(self, data: MyDataType):
      pass

    def calculate_antinodes(self, point1: Point, point2: Point, data: MyDataType) -> set[Point]:
      delta_row = point2[0] - point1[0]
      delta_col = point2[1] - point1[1]
      ret = set[Point]()
      for pos in [
        (point1[0] - delta_row, point1[1] - delta_col),
        (point2[0] + delta_row, point2[1] + delta_col),
      ]:
        if pos[0] in range(data.num_rows) and pos[1] in range(data.num_cols):
          ret.add(pos)
      return ret


puzzle: Puzzle = Puzzle()
puzzle.main()